### ISSUE TITLE
How are the gradients implemented for non-full rank matrices?

### ISSUE THREAD

**Issue by lezcano** [toshas/torch-householder#1]
It is not clear how to implement the "gradient" (adjoint) of the QR decomposition for a matrix that it is not full rank (e.g. the zero matrix). How does this package handle this?

Also, if this is a drop-in replacement for the QR decomposition implemented in PyTorch and it works better, why not making a PR to core PyTorch with this?  Where does the speed-up come from vs LAPACK  / MAGMA?

**Comment by toshas** [comment]
Hi @Lezcano, nice e-meeting you,
_I was keeping an eye on your effort with `mexp` - great job!_

> It is not clear how to implement the "gradient" (adjoint) of the QR decomposition for a matrix that it is not full rank (e.g. the zero matrix). How does this package handle this?

Check out the 5-liner reference implementation of the algorithm: https://github.com/toshas/torch-householder/blob/master/tests/reference_orgqr.py#L16-L20

Point being, Householder reflectors themselves are full-rank, and hence there is no problem with rank-deficiency. What it is multiplied by in the end is the truncated identity matrix (of which we don't need the gradients unless we compute `ormqr`).

> Also, if this is a drop-in replacement for the QR decomposition implemented in PyTorch and it works better, why not making a PR to core PyTorch with this? 

Great question! A small correction, this is not a QR decomposition, but quite the inverse -- a parameterized (orthogonal) matrix A is computed from its QR decomposition, where Q is parameterized by Householder reflectors, and R is the truncated identity.

For the second part of your question, contributing to opensource is definitely on the agenda. For now this package is the first step, also necessary to facilitating further projects.

> Where does the speed-up come from vs LAPACK / MAGMA?

Did not measure against those; last I checked there was no GPU support for `torch.orgqr`. For the efficiency in general, it comes part from the efficient Householder reflector-matrix product computation, part from C++ (avoid popping back into Python `r` times per matrix), part from GPU vectorization and parallelism offset to batch dimension.

Anton

**Comment by lezcano** [comment]
Good to e-meet you too, Anton!

Oh, sorry for the naïve questions from the first post. I now have read through the whole thing a bit more in detail and I think I understand it better. Just to check that my understanding is correct, what this does is, it takes a matrix of size `n x k`, with columns representing `k` vectors `v_i`, and if we denote `H(v) = I_n - v v^T / \norm{v}^2` a Householder reflection, what this does is, it computes `\prod_{i=1}^k H(v_i)`, is that right?

If that is the case, that sounds like a pretty nice thing since, as you have already mentioned, it is not necessary to compute a whole QR decomposition to parametrise `St(n,k)`. This is particularly helpful, as the QR decomposition is "not differentiable" on singular matrices (it can be made differentiable via QR with pivoting, which I have been meaning to include in PyTorch for sometime now, but that's another story).

For what is worth, I did not know that the product of Householder reflections of vectors of the form `lower_triangular + I_n` spanned the whole `St(n,k)`. Where can I find this result? (It might very well be the case that it follows trivially from looking at the Householder construction of the QR, but dunno).

In the case that the parametrisation is surjective, that means that the parametrisation cannot be a submersion everywhere (surjective differential) so, if this is the case, it would be interesting to look where does it fail to be a submersion.

Another minor thing is that, when `k = n`, when `n` is even, this parametrisation parametrises `SO(n)`, while when it is odd, it parametrises `O^-(n)` (the other coset) which is slightly annoying, but does not really matter for most applications.

Sorry for all the questions and comments, I just think that, at first sight, this could be quite an interesting approach to optimisation with orthogonal constraints :)

Edit. Forgot to mention about MAGMA. By the looks of it, MAGMA does have an `orgqr` implementation:
https://www.icl.utk.edu/~mgates3/docs/magma-proto.html
it would be interesting to see how well it performs, as it may even improve this (already very performant) implementation of yours.

**Comment by lezcano** [comment]
For what is worth, I think it would be best to write the benchmark with the benchmarking framework from PyTorch, although I reckon that it won't change too much, but just to be sure. See
https://pytorch.org/tutorials/recipes/recipes/benchmark.html

**Comment by toshas** [comment]
> \prod_{i=1}^k H(v_i), is that right?

Right.

> the product of Householder reflections of vectors of the form lower_triangular + I_n spanned the whole St(n,k). Where can I find this result?

That's how `geqrf` works, the following paper can serve as a reference:
https://pubs.acs.org/doi/pdf/10.1021/acs.jpca.5b02015

> Another minor thing is that, when k = n, when n is even, this parametrisation parametrises SO(n), while when it is odd, it parametrises O^-(n) (the other coset) which is slightly annoying, but does not really matter for most applications.

True, but if one needs SO(n) with `n = k = 2p+1`, it suffices to take R = diag(1,...,1,-1) or any other such convention.

> at first sight, this could be quite an interesting approach to optimisation with orthogonal constraints :)

It is indeed, I'd be very curious to play with the parameterization framework, and especially when magma orgqr is hooked up to pytorch!

> benchmarking framework from PyTorch

I glanced through it, had I known maybe would use it, but I seem to follow all recommendations in the part of warmup/etc.

**Comment by toshas** [comment]
PyTorch `orgqr` feature request: https://github.com/pytorch/pytorch/issues/50104

**Comment by lezcano** [comment]
I have a small script I use to benchmark things. If I find time tomorrow, I'll write a short PR with a benchmark  to this repo, in case you want to use it :)

You are totally right about the R = diag. That's definitely the way to go.

And I see, you have that tau computation behind the scenes. It is there were you find the critical points of this parametrisation. This parametrisation has a zero gradient when the parameters are zero. That can definitely bite people, but well, it is what it is. In general, it is not full rank whenever any of the vectors `v_i` is zero. It is slightly annoying, but on the flip side, it is very easy to detect when that is the case, which is quite nice.

**Comment by lezcano** [comment]
Also, thanks for the feature request. Perhaps when I finish writing the thesis in a few weeks I'll find time to do a PR to PyTorch adding support for orgqr. We'll see...

**Comment by toshas** [comment]
> And I see, you have that tau computation behind the scenes. It is there were you find the critical points of this parametrisation. This parametrisation has a zero gradient when the parameters are zero. That can definitely bite people, but well, it is what it is. In general, it is not full rank whenever any of the vectors v_i is zero.

I'm curious to understand the limitations of the proposed parameterization, but in this case I'm not sure I follow why the gradients of parameters (located under the diagonal) are zero when parameters are zero. Here is a simple test case using the [reference function](https://github.com/toshas/torch-householder/blob/master/tests/reference_orgqr.py#L4):

```
>>> param = torch.nn.Parameter(torch.eye(4, 2))
>>> tau = 2 / (param * param).sum(dim=0)
>>> q = reference_orgqr(param, tau)
>>> synthetic_loss = (q * torch.arange(8).view(4,2).float()).sum()
>>> synthetic_loss.backward()
>>> print(param.grad)
tensor([[  0.,  -2.],
        [ -2.,   0.],
        [ -8., -10.],
        [-12., -14.]])
```

Granted, I omitted the `tril(diagonal=-1)` mask, however the gradients look perfectly fine (non-zero) to me. 

For the `tau` argument of `orgqr`, this is an artifact of the LAPACK API, which I presume was introduced to avoid having to take the inverse sqrt. In the [tests suite](https://github.com/toshas/torch-householder/blob/master/tests/test.py) I check different cases when [tau is supplied, or not supplied] x [tau is a independent parameter, or an intermediate variable].

> Perhaps when I finish writing the thesis in a few weeks I'll find time to do a PR to PyTorch adding support for orgqr.

That'd be great; otherwise we can wait for community managers to prioritize the issue. 
Good luck with the writing!

**Comment by lezcano** [comment]
Hi again!

About the differential, you are right. I did that computation of the top of my head and it was clearly wrong. I do not have the time now to do it right now, but I will take a look at it whenever I have a bit more time.

Also, here's a short example of a benchmark using `torch.utils.benchmark`, in case you want to use it. I reckon it'll be more reliable, but I don't expect to see huge differences with respect to the current times.

Also, feel free to close this issue, I think I understand much better now the point of this package. In any case, we can move any discussion to the PyTorch issue that you opened :)

```python
import itertools as it
import sys
import time

import numpy as np
import torch
import torch.utils.benchmark as benchmark_utils

np.random.seed(11)
torch.manual_seed(13)

MAX_REPLICATES = 5
PARAMS = list(it.product(
    [16, 32, 64, 128],  # Batch
    [64, 256, 1024],    # Size
))

def run_test(input_generator, stmt_generator, expm=True):
    start_time = time.time()
    results = []
    for batch, size in PARAMS:
        cpu_threads = [4, 16]

        x = np.random.rand(batch, size, size) / size
        if expm:
            a = input_generator(batch, size)
            stmt = stmt_generator("a")
            lbl = "Matrix Exponential"
            gl = {"a": a}
        else:
            a,b = input_generator(batch, size)
            stmt = stmt_generator("a", "b")
            lbl = "Triu Solver"
            gl = {"a": a, "b":b}

        #for num_threads in cpu_threads:
        timer = benchmark_utils.Timer(
            stmt=stmt,
            globals=gl,
            label=lbl,
            sub_label=f"batch: {batch}",
            description=f"{size}x{size}",
            num_threads=torch.get_num_threads(),
        )

        total_count = 0
        for replicate in range(MAX_REPLICATES):
            # If the run-to-run variation is too high, a warning is associated
            # with each measurement. In such a case, do extra replicates. These
            # will be combined in post processing.
            if replicate == 0 or results[-1].has_warnings or total_count < 5:
                r = timer.blocked_autorange(min_run_time=5)
                total_count += len(r.times)
                results.append(r)
                print(".", end="")
                sys.stdout.flush()

    print()
    sys.stdout.flush()

    comparison = benchmark_utils.Compare(results)
    comparison.trim_significant_figures()
    comparison.print()
    print(f"Benchmark time: {time.time() - start_time:.1f} sec")

def main():
    pytorch_input_gen_gpu = lambda batch, size: torch.rand(batch, size, size).cuda() / size

    def pytorch_triu_solve(batch, size):
        return torch.rand(batch, size, size).triu().cuda() / size, torch.rand(batch, size, size).cuda() / size

    print("PyTorch Triu")
    run_test(
        pytorch_triu_solve,
        lambda A, B: "torch.triangular_solve({}, {})".format(B, A),
        expm=False
    )

    print("PyTorch matrix_exp")
    run_test(
        pytorch_input_gen_gpu,
        lambda var_name: var_name + ".matrix_exp()"
    )


if __name__ == "__main__":
    main()
```
